iac-common/src/main/java/org/sonar/iac/common/yaml/visitors/YamlMetricsVisitor.java

```diff
@@ -35,26 +36,33 @@ public YamlMetricsVisitor(FileLinesContextFactory fileLinesContextFactory, NoSon
   }
 
   @Override
-  protected void languageSpecificMetrics() {
-    register(ScalarTree.class, (ctx, tree) -> {
-      for (int i = tree.textRange().start().line(); i <= endLine(tree); i++) {
-        linesOfCode().add(i);
-      }
-    });
-    register(MappingTree.class, (ctx, tree) -> addBraces(tree));
-    register(SequenceTree.class, (ctx, tree) -> addBraces(tree));
-    register(YamlTree.class, (ctx, tree) -> addCommentLines(tree.metadata().comments()));
+  protected void before(InputFileContext ctx, Tree root) {
+    super.before(ctx, root);
+    try {
+      analyzeFileContentForLoc(ctx.inputFile);
+    } catch (IOException e) {
+      throw new ParseException("Can not read file for metric calculation", null, e.getMessage());
+    }
   }
 
-  // SONARIAC-82 Lines which contain only brackets should also be counted for metrics
-  private void addBraces(HasTextRange tree) {
-    linesOfCode().add(tree.textRange().start().line());
-    linesOfCode().add(endLine(tree));
+  /**
+   * Analyze line by line and do not use nodes from YAML parser since their text range contains trailing lines with whitespaces which should
+   * not be counted as LOC
+   */
+  private void analyzeFileContentForLoc(InputFile inputFile) throws IOException {
+    int lineNumber = 1;
+    String[] lines = inputFile.contents().split("(\\r\\n|\\r|\\n)");
+    for (String line : lines) {
```

Comment:
Would iterating over lines without reading the entire file into memory make sense here? Or will the difference be negligible, because we read it later anyway for parsing?

Reply:
Thank you for the suggestion. I was wondering if `InputFile` provides a proper API for this, but I was overlooking the `InputFile::inputStream` method.